fathom/src/core/binary.rs

```diff
@@ -208,6 +193,79 @@ impl<'arena, 'env> Context<'arena, 'env> {
 
         Ok(Arc::new(Value::Const(Const::Ref(pos))))
     }
+
+    fn read_deref(
+        &mut self,
+        reader: &mut dyn SeekRead,
+        format: &ArcValue<'arena>,
+        r#ref: &ArcValue<'arena>,
+    ) -> io::Result<ArcValue<'arena>> {
+        let pos = match self.elim_context().force(r#ref).as_ref() {
+            Value::Const(Const::Ref(pos)) => *pos,
+            _ => {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "invalid format reference",
+                ))
+            }
+        };
+
+        self.read_cached_ref(reader, pos, format)
+    }
+
+    fn lookup_cached_ref<'context>(
+        &'context self,
+        pos: u64,
+        format: &ArcValue<'_>,
+    ) -> Option<&'context ParsedRef<'arena>> {
+        // NOTE: The number of calls to `semantics::ConversionContext::is_equal`
+        // when looking up cached references is a bit of a pain. If this ever
+        // becomes a problem we could improve performance by pre-allocating a
+        // `ParsedRef` in the cache during `read_link`, and storing the index of
+        // that parsed reference alongside the position in `Const::Ref`.
+
+        (self.cached_refs.get(&pos)?.iter())
+            .find(|r| self.conversion_context().is_equal(&r.format, &format))
+    }
+
+    fn read_cached_ref(
+        &mut self,
+        reader: &mut dyn SeekRead,
+        pos: u64,
+        format: &ArcValue<'arena>,
+    ) -> io::Result<ArcValue<'arena>> {
+        if let Some(parsed_ref) = self.lookup_cached_ref(pos, &format) {
+            return Ok(parsed_ref.expr.clone());
+        }
+
+        let initial_pos = reader.stream_position()?;
+
+        // Seek to current current ref location
+        reader.seek(SeekFrom::Start(pos))?;
+        // Parse the data at that location
+        let expr = self.read_format(reader, &format)?;
+        // Reset reader back to the original position
+        reader.seek(SeekFrom::Start(initial_pos))?;
+
+        // We might have parsed the current reference during the above call to
+        // `read_format` so it's important to check first, in order to avoid
+        // duplicates in the cache.
+        if let Some(parsed_ref) = self.lookup_cached_ref(pos, &format) {
```

Comment:
I think this should be an assertion failure as it should not ever be possible for this to happen as it would imply infinite regress.